### ISSUE TITLE
Incorrect mode selected with direct mini boot

### ISSUE THREAD

**Issue by neagix** [neagix/wii-linux-ngx#2]
Starting from a completely shutdown Wii, with zImage to boot via default through mini:
* incorrect mode is selected (780i) on a PAL TV set
* black screen, no console can be seen

Starting from a completely shutdown Wii, with Bootmii shell GUI selected to boot via default through mini:
* 780p is selected by Bootmii
* load Linux from the SD
* the console will now always work, even across reboots (!!!)

It took a while to figure this out, but it's reproducible. The bug must be in the `tv=auto` detection, combined with something that stays alive in GPU across reboots.

**Comment by DeltaResero** [comment]
It seems almost like the Wii isn't shutting down properly. Not sure if it's related or not, but I've noticed that when I boot into MINI from BootMii with the rebased patches I did and then restart, when I get back to the Wii Warning Heath and Safety Screen, I'm unable to sync my Wii Remotes normally forcing me to use the red sync bottom on both console and remote.

**Comment by neagix** [comment]
@DeltaResero oh yes, I noticed that too. It's happening every time, I didn't know it's a new bug from v2.6 kernels.

And that's also why I have set Bootmii to be booted directly through Priiloader, to avoid having to use the Wiimote.

**Comment by neagix** [comment]
Rather than not proper shutdown, could it be that the memory where bluetooth info is stored (perhaps in Starlet?) get corrupted?

I am a Wii internals noob, so that's just a wild guess from me.

**Comment by neagix** [comment]
FYI, I have tried this commit 1ebe29cc73391ed4c5dc627dff9e50eedbb9f7ab from Herranz, in case it would add the necessary initialization, but no luck.

It does break the ttyUSB support for some reason. Branch: https://github.com/neagix/wii-linux-ngx/tree/broken-v3.x-herranz (I might revert it and add different commits there though)

**Comment by neagix** [comment]
I am thinking that maybe the kernel changed the way it does I2C initialization and now it happens somehow after the framebuffer, so the framebuffer doesn't have the I2C client to the AVE chipset?

This should be relatively easy to verify.

**Comment by neagix** [comment]
@DeltaResero as you mentioned the cf imageblit speedup commit (ce63afebd79db8b0b7c085387311814e668c4bd7), I can confirm it does not affect this problem or the issues remaining in #6. I still think it's some I2C race condition e.g. frame buffer initialised before AVE is initialised

**Comment by neagix** [comment]
I added the kernel message; apparently it fails to request the memory, very strange.

**Comment by neagix** [comment]
I think the problem is two-fold:
* setting an invalid TV mode has no effect, leaving untouched the mode which is already there
* AVE is not probed (via I2C), so perhaps this is the regression from v2.6.x

**Comment by marcan** [comment]
If you need anything from the BootMii UI just tag me and ask, though it seems the main issue is on the Linux side here. Obviously if you don't talk to AVE you won't get any video, especially on cold boot.

**Comment by neagix** [comment]
@marcan thanks for commenting, didn't know you were on GH! Yes I am narrowing down the problem and basically I believe the `wii.dts` file is not in good shape.

I think AVE stays initialized across a CPU reset, and that's why things work after hopping through the BootMii UI.

Look here:
```
/hollywood nintendo,hollywood
/hollywood/audio@0c005000 nintendo,hollywood-audio
/hollywood/control@0d800100 nintendo,hollywood-control
/hollywood/disk@0d806000 nintendo,hollywood-di
/hollywood/exi@0d006800 nintendo,hollywood-exinintendo,flipper-exi
/hollywood/gamepad-controller@0d006400 nintendo,hollywood-sinintendo,flipper-si
/hollywood/ios@0d000000 nintendo,starlet-ios-ipc
/hollywood/mini@0d000000 twiizers,starlet-mini-ipc
/hollywood/mini@0d000000/gpio-leds gpio-leds
/hollywood/mini@0d000000/gpio@0d8000c0 nintendo,hollywood-gpio
/hollywood/mini@0d000000/i2c-video virtual,i2c-gpio
/hollywood/mini@0d000000/pic1@0d800030 nintendo,hollywood-pic
/hollywood/mini@0d000000/sd@0d070000 nintendo,hollywood-sdhcisdhci
/hollywood/mini@0d000000/sdio@0d080000 nintendo,hollywood-sdhcisdhci
/hollywood/mini@0d000000/usb@0d040000 nintendo,hollywood-usb-ehciusb-ehci
/hollywood/mini@0d000000/usb@0d050000 nintendo,hollywood-usb-ohciusb-ohci
/hollywood/mini@0d000000/usb@0d060000 nintendo,hollywood-usb-ohciusb-ohci
/hollywood/pic0@0c003000 nintendo,flipper-pic
/hollywood/resetswitch@0c003000 nintendo,hollywood-pinintendo,flipper-pi
/hollywood/video@0c002000 nintendo,hollywood-vinintendo,flipper-vi
```

`hollywood-pinintendo` and `hollywood-vinintendo` don't seem right. At the moment I have no idea how this is happening..but basically since there is no device in the tree with matching string `wii-ave-rvl`, AVE never gets probed.

**Comment by marcan** [comment]
Where are you getting that dump from? That looks like it's concatenating multiple 'compatible' values. The real DTS looks like:
```
			compatible = "nintendo,hollywood-pi",
					"nintendo,flipper-pi";
```

**Comment by marcan** [comment]
AVE is supposed to be nested under /hollywood/mini@0d000000/i2c-video FWIW

**Comment by neagix** [comment]
@marcan yeah, the device tree in procfs uses \0 as separator, that was why. I have edited the comment.

Yes, AVE is there in that path. It is really not getting a driver though (if I am not interpreting wrong):
```sh
# find /sys/devices/ -name uevent -exec grep -H video \{\} \;
/sys/devices/hollywood.0/d000000.mini/i2c-video.2/uevent:OF_NAME=i2c-video
/sys/devices/hollywood.0/d000000.mini/i2c-video.2/uevent:OF_FULLNAME=/hollywood/mini@0d000000/i2c-video
/sys/devices/hollywood.0/d000000.mini/i2c-video.2/uevent:MODALIAS=of:Ni2c-videoT<NULL>Cvirtual,i2c-gpio
/sys/devices/hollywood.0/c002000.video/uevent:OF_NAME=video
/sys/devices/hollywood.0/c002000.video/uevent:OF_FULLNAME=/hollywood/video@0c002000
/sys/devices/hollywood.0/c002000.video/uevent:MODALIAS=of:NvideoT<NULL>Cnintendo,hollywood-viCnintendo,flipper-vi
```

**Comment by neagix** [comment]
There is no `/dev/i2c-0` and that explains why the AVE gets no device and no driver.

Regarding i2c, I put all my suspects on the `i2c-gpio-of` but I am really not that good to debug this. It's either a regression on the patches by @DeltaResero from v2.6 to v3.x, or a bug introduced through other commits by other kernel developers in the v3.x tree.

By the way, the only nodes without a device are those of IOS (makes sense when booting in MINI mode) and this one for the AVE, which instead should be there.

By reading around I have found that we could in theory change it to be a more explicit initialisation: http://elixir.free-electrons.com/linux/v3.15.10/source/Documentation/i2c/instantiating-devices

However, changing this now resembles more "shotgun debugging" than rather finding out why i2c simply stopped working.

Elsewhere I've read that the I2C might simply be not in the correct state, but I also want to exclude this as I believe MINI does all the correct initialisation (since it works with v2.6).

I have also checked the patches by @DeltaResero and they seem fine, the DTS specifies `reg = <0x0d8000c0 0x40>` and GPIOs 16,17
